### PR TITLE
Confirm we send SNI server names on the JDK9 Platform

### PR DESCRIPTION
**Pull request by swankjesse**
This introduces a new API in MockWebServer's RecordedRequest to capture the inbound server names.

### CODE REVIEW

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt [hunk 1]
```diff
@@ -37,6 +38,10 @@ open class Jdk9Platform : Platform() {
 
     sslParameters.applicationProtocols = names.toTypedArray()
 
+    if (hostname != null) {
+      sslParameters.serverNames = listOf(SNIHostName(hostname))
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt:42]: My read is we haven’t done SNI correctly on non-Java platforms. . . . forever? Wow!

Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt:42]: My read is we haven’t done SNI correctly on non-Android platforms. . . . forever? Wow!

Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt:42]: It was deliberate. Its meant to be inbuilt

Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt:42]: Maybe our usage of the Socket API breaks this?

```
        // Create the wrapper over the connected socket.
        val sslSocket = route.address.sslSocketFactory.createSocket(
          rawSocket,
          route.address.url.host,
          route.address.url.port,
          true /* autoClose */
        ) as SSLSocket
```        
        
Surely the host here should signal SNI, which was enable for clients by default in JDK 7?

File: mockwebserver/src/test/java/mockwebserver3/MockResponseSniTest.kt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    val recordedRequest = server.takeRequest()
+    assertThat(recordedRequest.handshakeServerNames).containsExactly(url.host)
+  }
+
+  /**
+   * Use different hostnames for the TLS handshake (including SNI) and the HTTP request (in the
+   * Host header).
+   */
+  @Test
+  fun domainFronting() {
```
Comment (yschimke) [inline mockwebserver/src/test/java/mockwebserver3/MockResponseSniTest.kt:69]: Nice!

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/AndroidPlatform.kt [hunk 3]
```diff
@@ -39,7 +39,7 @@ import okhttp3.internal.tls.BasicTrustRootIndex
 import okhttp3.internal.tls.CertificateChainCleaner
 import okhttp3.internal.tls.TrustRootIndex
 
-/** Android 5+. */
+/** Android 5+ (API 21+). */
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/AndroidPlatform.kt:42]:
```suggestion
/** Android 5 to 9 (API 21 to 28). */
```

**Review by yschimke — APPROVED** [review]
Good catch, although I'm going to drill in, work out why the automatic SNI functionality of the JDK isn't working.

**Comment by yschimke** [comment]
Failure

```
CallTest[jvm] > ipv6HostHasSquareBracesHttp1()[jvm] FAILED
    java.lang.IllegalArgumentException: Contains non-LDH ASCII characters
        at java.base/java.net.IDN.toASCIIInternal(IDN.java:296)
        at java.base/java.net.IDN.toASCII(IDN.java:122)
        at java.base/javax.net.ssl.SNIHostName.<init>(SNIHostName.java:99)
        at okhttp3.internal.platform.Jdk9Platform.configureTlsExtensions(Jdk9Platform.kt:42)
        at okhttp3.internal.connection.ConnectPlan.connectTls(ConnectPlan.kt:322)
        at okhttp3.internal.connection.ConnectPlan.connectTlsEtc(ConnectPlan.kt:191)
        at okhttp3.internal.connection.FastFallbackExchangeFinder.find(FastFallbackExchangeFinder.kt:78)
        at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:267)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:84)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:65)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:205)
        at okhttp3.internal.connection.RealCall.execute(RealCall.kt:158)
        at okhttp3.CallTest.ipv6HostHasSquareBracesHttp1(CallTest.kt:3692)
```

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt:42]: I think it doesn’t, and I’m horrified. In particular I think we needed to manually opt-in with the `SNIHostName` thing.

**Comment by yschimke** [comment]
I've merged master.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt (hunk 1 above)
Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt:43]: OK, so I don't think it's this bad.  I think we don't understand the full set of things going on.  Your test ends up with "127.0.0.1" 

image

While with an external site it's the correct hostname

image

Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt:43]: localhost is not a valid SNI name.

image

**Comment by yschimke** [comment]
@swankjesse I've reverted setting the SNI names, and just updated the tests.  Take a look.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt (hunk 1 above)
Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Jdk9Platform.kt]: Spec: https://datatracker.ietf.org/doc/html/rfc6066#page-6

I assume it's this

>  "HostName" contains the fully qualified DNS hostname of the server, as understood by the client.

File: mockwebserver/src/test/java/mockwebserver3/MockResponseSniTest.kt [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  private lateinit var server: MockWebServer
+
+  @BeforeEach
+  fun setUp(server: MockWebServer) {
+    this.server = server
+  }
+
+  @Test
+  fun clientSendsServerNameAndServerReceivesIt() {
+    // java.net.ConnectException: Connection refused
```
Comment (yschimke) [inline mockwebserver/src/test/java/mockwebserver3/MockResponseSniTest.kt:48]: I should look into this, but this test is better than no test, as it confirms our current behaviour.

**Comment by yschimke** [comment]
Windows failures are unrelated.

**Pull request merged**